File: lc/lc_q1081.cpp
#include <bits/stdc++.h>
class Solution {
    public :
        string smallestSubsequence(string s){
            int freq[26]={0};
            bool seen[26]={0};
            stack <int> stk;
            for (int i=0;i<s.size();i++){
                freq[s[i]-'a']++;
            }
            for (auto c : s){
                freq[c-'a']--;
                if (seen[c-'a']==1){
                    continue;
                }
                while(!stk.empty()&&c<stk.top()&&freq[stk.top()-'a']>0){
                    seen[stk.top()-'a']=0;
                    stk.pop();
                }
                stk.push(c);
                seen[c-'a']=1;
            }
            string ans;
            while(!stk.empty()){
                ans+=stk.top();
                stk.pop();
            }
            reverse(ans.begin(),ans.end());
            return ans;
        }

};

File: playground/lc_q1081.java
import java.util.*;
class Solution {
    public string smallestSubsequence(string s){
        int[] freq=new int[26];
        boolean[] seen=new boolean[26];
        Stack <Character> stack = new Stack<>();
        for (int i=0;i<s.length();i++){
            freq[s.charAt(i)-'a']++;
        }
        for (char c:s.toCharArray()){
            freq[c-'a']--;
            if (seen[c-'a']){
                continue;
            }
            while(!stack.isEmpty()&&stack.peek()>c&&freq[stack.peek()-'a']>0){
                seen[stack.pop()-'a']=false;
            }
            stack.push(c);
            seen[c-'a']=true;
        }
        StringBuilder ans=new StringBuilder();
        for (char c:stack){
            ans.append(c);
        }
        return ans.toString();
    }
}

